### PR TITLE
[dynamo] Graph break on `setattr(Tensor, "data", Tensor)`

### LINKED ISSUE #113030: Mismatching behaviour of tensor assignment "a.data = b" between torch.compile and eager execution
### 🐛 Describe the bug

Since Pytorch 2.1, execution of functions with "a.data = b" tensor assignment produces different results than eager execution. 
In PT2.0 eager and compile and PT2.1 eager mode, tensorA = tensorB is treated like tensorA = torch.clone(tensorB), whereas in compile since PT2.1, it executes like it was tensorA = tensorB.view(-1).

Here's a code snippet confirming this behavior:
```python
import torch

print("input tensors a, b:")
a = torch.rand([6])
a1 = torch.clone(a)
print(a)
b = torch.rand([6])
b1 = torch.clone(b)
print(b)

def func(x,y):
  x.data = y
  y.data = torch.zeros(0)
  return x

f_compiled = torch.compile(func)
f_compiled(a,b)
print("\n a, b after compile run:")
print(a)
print(b)

func(a1,b1)
print("\n a, b after eager run:")
print(a1)
print(b1) 
```
PT2.1 and 2.2 result:

```python
input tensors a, b:
tensor([0.2961, 0.5166, 0.2517, 0.6886, 0.0740, 0.8665])
tensor([0.1366, 0.1025, 0.1841, 0.7264, 0.3153, 0.6871])

a, b after compile run:
tensor([])
tensor([])

a, b after eager run:
tensor([0.1366, 0.1025, 0.1841, 0.7264, 0.3153, 0.6871])
tensor([])
```
PT2.0 result:

```python
input tensors a, b:
tensor([0.1951, 0.3202, 0.4022, 0.1943, 0.7430, 0.9812])
tensor([0.1001, 0.1426, 0.0979, 0.7146, 0.9821, 0.6137])

a, b after compile run:
tensor([0.1001, 0.1426, 0.0979, 0.7146, 0.9821, 0.6137])
tensor([])

a, b after eager run:
tensor([0.1001, 0.1426, 0.0979, 0.7146, 0.9821, 0.6137])
tensor([]) 
```
Execution results do not differ between used devices (CPU and HPU).

### Error logs

_No response_

### Minified repro

_No response_

### Versions

PyTorch version: 2.2.0.dev20231106+cpu
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.3 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: version 3.26.1
Libc version: glibc-2.35

Python version: 3.10.12 (main, Jun 11 2023, 05:26:28) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-5.15.90.1-microsoft-standard-WSL2-x86_64-with-glibc2.35
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Address sizes:                   39 bits physical, 48 bits virtual
Byte Order:                      Little Endian
CPU(s):                          8
On-line CPU(s) list:             0-7
Vendor ID:                       GenuineIntel
Model name:                      11th Gen Intel(R) Core(TM) i7-1185G7 @ 3.00GHz
CPU family:                      6
Model:                           140
Thread(s) per core:              2
Core(s) per socket:              4
Socket(s):                       1
Stepping:                        1
BogoMIPS:                        3609.59
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon rep_good nopl xtopology tsc_reliable nonstop_tsc cpuid pni pclmulqdq vmx ssse3 fma cx16 pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves avx512vbmi umip avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq rdpid movdiri movdir64b fsrm avx512_vp2intersect flush_l1d arch_capabilities
Virtualization:                  VT-x
Hypervisor vendor:               Microsoft
Virtualization type:             full
L1d cache:                       192 KiB (4 instances)
L1i cache:                       128 KiB (4 instances)
L2 cache:                        5 MiB (4 instances)
L3 cache:                        12 MiB (1 instance)
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Not affected
Vulnerability Retbleed:          Mitigation; Enhanced IBRS
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Enhanced IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected

Versions of relevant libraries:
[pip3] numpy==1.24.1
[pip3] torch==2.2.0.dev20231106+cpu
[pip3] torchaudio==2.2.0.dev20231106+cpu
[pip3] torchvision==0.17.0.dev20231106+cpu
[pip3] triton==2.0.0
[conda] Could not collect

cc @ezyang @msaroufim @wconstab @bdhirsh @anijain2305 @zou3519

### PR DESCRIPTION
**Pull request by jon-chuang**
Fixes https://github.com/pytorch/pytorch/issues/113030

Alias information needs to be applied in eager before we can continue to trace the graph.

----

Perhaps this is too strict - couldn't we fx trace through the in-graph (pointer) aliasing, and track mutations through fake tensors instead, and still apply the aliasing mutation epilogue for further mutations outside of graph? :thinking: 

Regardless, it didn't seem to work too well when I tried this. Seems that `Tensor.__setattr__` doesn't work well in fx graph.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

File: torch/_dynamo/variables/builtin.py [hunk 1]
```diff
@@ -1160,6 +1160,7 @@ def call_setattr(
     ):
         from .distributed import PlacementVariable
 
+        print("SETATTR", obj, tx.output.side_effects.is_attribute_mutation(obj))
```
Comment (ezyang) [inline torch/_dynamo/variables/builtin.py:1163]: print

**Review by ezyang — APPROVED** [review]
Thanks, appreciated. @voznesenskym, heads up!

**Comment by voznesenskym** [comment]
I can also just add support for it.

**Review by voznesenskym — CHANGES_REQUESTED** [review]
I think setting .data = None will fall through here?

**Comment by jon-chuang** [comment]
Good point, let's just forbid setattr data generally. 

> I can also just add support for it.

@albanD also mentioned in slack that it's p hard to support .data and we're trying to migrate used off it.

**Comment by jon-chuang** [comment]
> Good point, let's just forbid setattr data generally.

Actually, Dynamo will handle this case. `TypeError: Variable data has to be a tensor, but got NoneType`

**Comment by jon-chuang** [comment]
> Actually, Dynamo will handle this case. TypeError: Variable data has to be a tensor, but got NoneType

I added this behaviour as a test case

**Comment by voznesenskym** [comment]
Actually, let me just add a PR to enable this, I am almost done.

**Comment by voznesenskym** [comment]
> > Good point, let's just forbid setattr data generally.
> 
> Actually, Dynamo will handle this case. `TypeError: Variable data has to be a tensor, but got NoneType`

Yeah, you're right :)

**Comment by voznesenskym** [comment]
https://github.com/pytorch/pytorch/pull/113069

**Review by voznesenskym — APPROVED** [review]
LGTM

**Comment by jon-chuang** [comment]
@voznesenskym I'm gonna merge this first then you can adapt the tests for https://github.com/pytorch/pytorch/pull/113069/? Maybe could add a fullgraph=True and change frame_count = 1

**Comment by voznesenskym** [comment]
I'll be reverting this in my grad enablement PR

**Comment by voznesenskym** [comment]
> @voznesenskym I'm gonna merge this first then you can adapt the tests for #113069?
> 
> In that PR you could add a fullgraph=True and change frame_count = 1

That's already the case

**Comment by PaliC** [comment]
@pytorchbot revert -m "causing dynamo regressions internally (see error message below)" -c ghfirst

```
self = <caffe2.test.dynamo.test_input_attr_tracking.TestInputAttrTracking testMethod=test_set_data_on_input_tensor>

    def test_set_data_on_input_tensor(self):
        def fn(x, y):
            x.data = y.data
            if x.size() == y.size():
                return x * y
            else:
                return y * y
    
        x = torch.randn([5, 5])
        y = torch.randn([2, 2])
    
        eager_result = fn(x, y)
    
        counter = CompileCounter()
    
        fn = torch._dynamo.optimize(counter, nopython=True)(fn)
    
>       compile_result = fn(x, y)

...

msg = '.data assignment to a tracked tensors can introduce aliasing, hence we need to graph break to apply the aliasing (or track new aliased tensors) to continue to trace the graph'

    def unimplemented(msg: str) -> NoReturn:
        assert msg != os.environ.get("BREAK", False)
>       raise Unsupported(msg)
E       torch._dynamo.exc.Unsupported: .data assignment to a tracked tensors can introduce aliasing, hence we need to graph break to apply the aliasing (or track new aliased tensors) to continue to trace the graph
E       
E       from user code:
E          File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/buck-out/v2/gen/fbcode/6081734815403318/caffe2/test/dynamo/__test_dynamo__/test_dynamo#link-tree/caffe2/test/dynamo/test_input_attr_tracking.py", line 282, in fn
E           x.data = y.data
E       
E       Set TORCH_LOGS="+dynamo" and TORCHDYNAMO_VERBOSE=1 for more information
E       
E       
E       You can suppress this exception and fall back to eager by setting:
E           import torch._dynamo
E           torch._dynamo.config.suppress_errors = True

...

[internal_folder]/caffe2/test/dynamo/__test_dynamo__/test_dynamo#link-tree/torch/_dynamo/exc.py:192: Unsupported
"
                            [resultArtifactsHandle:TestInfraTestResult:private] => 
                            [metadata:TestInfraTestResult:private] => Array
                                (
                                )

                            [metadataLoaded:TestInfraTestResult:private] => 1
                            [metadataHandle:TestInfraTestResult:private] => 
                            [prov:TestInfraTestResult:private] => TestInfraTestDataProvider Object
                                (
                                    [queryTimeout:TestInfraTestDataProvider:private] => 30
                                    [instanceKey:TestInfraTestDataProvider:private] => "1699462559.63675816"
                                    [shardedRun:TestInfraTestDataProvider:private] => TestInfraShardedRun Object
                                        (
                                            [queryTimeout:TestInfraShardedRun:private] => 30
                                            [readOnMaster:TestInfraShardedRun:private] => 1
                                            [caller:TestInfraShardedRun:private] => "SandcastleFBCodeContbuildBase"
                                        )

                                    [shardedTest:TestInfraTestDataProvider:private] => TestInfraShardedTest Object
                                        (
                                            [queryTimeout:TestInfraShardedTest:private] => 30
                                            [readOnMaster:TestInfraShardedTest:private] => 1
                                            [caller:TestInfraShardedTest:private] => "SandcastleFBCodeContbuildBase"
                                        )

                                    [readOnMaster:TestInfraTestDataProvider:private] => 1
                                    [caller:TestInfraTestDataProvider:private] => "SandcastleFBCodeContbuildBase"
                                )

                            [extraData:TestInfraTestResult:private] => Array
                                (
                                )

                            [verdictExplanation:TestInfraTestResult:private] => 
                            [completePromise:TestInfraTestResult:private] => 
                            [landFrictionOverride:TestInfraTestResult:private] => 
                            [suppressionInformation:TestInfraTestResult:private] => 
                        )

                    [arePreviousTriesLoaded:TestInfraTestResult:private] => 1
                    [summary:TestInfraTestResult:private] => ""
                    [details:TestInfraTestResult:private] => "============================= test session starts ==============================
platform linux -- Python 3.8.6, pytest-7.2.2, pluggy-0.13.1
rootdir: [redacted[
plugins: hypothesis-6.70.1
collected 12 items

. unimplemented [('.data assignment to a tracked tensors can introduce aliasing, hence we need to graph break to apply the aliasing (or track new aliased tensors) to continue to trace the graph', 1)]
F
```

Hopefully, this should be enough of an error message to help you debug

**Comment by jon-chuang** [comment]
I think the internal test needs to be xfailed until https://github.com/pytorch/pytorch/pull/113080 is merged. @voznesenskym 

Setting data is completely broken atm

**Comment by voznesenskym** [comment]
> I think the internal test needs to be xfailed until #113080 is merged. @voznesenskym
> 
> Setting data is completely broken atm

mmmm not clear, that test was passing. It should graph break now, this is not an internal test. Looking.

**Pull request closed without merging**